aptos-move/block-executor/src/scheduler_status.rs

```diff
@@ -0,0 +1,1387 @@
+    /// - `Ok(false)` if the transaction remains stalled after this call
+    /// - `Err` if there was an error removing the stall (e.g., no matching add_stall)
+    pub(crate) fn remove_stall(&self) -> Result<bool, PanicError> {
+        let prev_num_stalls = self.num_stalls.fetch_sub(1, Ordering::SeqCst);
+
+        if prev_num_stalls == 0 {
+            return Err(code_invariant_error(
+                "remove_stall called when num_stalls == 0",
+            ));
+        }
+
+        if prev_num_stalls == 1 {
+            // Acquire write lock for (non-monitor) shortcut modifications.
+            let inner_status = self.inner_status.lock();
+
+            // num_stalls updates are not under the lock, so need to re-check (otherwise
+            // a different add_stall might have already incremented the count).
+            if self.num_stalls.load(Ordering::Relaxed) > 0 {
+                return Ok(false);
+            }
+
+            if let Some(incarnation) = inner_status.pending_scheduling() {
+                if incarnation == 0 {
+                    // Invariant due to scheduler logic: for a successful remove_stall there
+                    // must have been an add_stall for incarnation 0, which is impossible.
+                    return Err(code_invariant_error("0-th incarnation in remove_stall"));
+                }
+                self.execution_queue_manager
+                    .add_to_schedule(incarnation == 1, self.txn_idx);
+            } else if inner_status.is_executed() {
+                // TODO(BlockSMTv2): Here, when waiting is supported, if inner status is executed,
+                // would need to notify waiting workers.
+
+                // Status is Executed so the dependency shortcut flag may not be
+                // EXECUTING (finish_execution sets Executed status and DEFER or SAFE flag).
+                self.swap_dependency_flag_any(
+                    &[dependency_flags::DEFER, dependency_flags::SAFE],
+                    dependency_flags::SAFE,
+                    "remove_stall",
+                )?;
+            }
+
+            return Ok(true);
+        }
+        Ok(false)
+    }
+
+    /// Checks if an incarnation has already been marked for abortion.
+    ///
+    /// This can be called during an ongoing execution to determine if the
+    /// execution has been concurrently aborted. This allows the executor
+    /// to return early and to discard the results.
+    pub(crate) fn already_try_aborted(&self, incarnation: Incarnation) -> bool {
+        self.next_incarnation_to_try_abort.load(Ordering::Relaxed) > incarnation
+    }
+
+    /// Checks that the shortcut flag is SAFE. This is used by the scheduler
+    /// when removing a previously propagated stall signal, and it is safe to
+    /// use the shortcutbecause of the best-effort nature of the stall mechanism.
+    pub(crate) fn shortcut_executed_and_not_stalled(&self) -> bool {
```

Comment:
Why not shortcut_executed_depndency_safe?

Reply:
I thought for the caller safe doesn't mean much, but inside the implementation it's a more concise term.